File: python/parser.py
from typing import List
from model import (
    Assignment,
    BinOp,
    Identifier,
    IfStatement,
    LogicalOp,
    Bool,
    Expression,
    Statement,
    Statements,
    String,
    UnaryOp,
    Float,
    Grouping,
    Integer,
    PrintStatement,
    PrintlnStatement,
    WhileStatement,
    ForStatement,
)
from tokens import Token, TokenType
from typing import Optional


class Parser:
    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.current = 0

    def advance(self) -> Optional[Token]:
        if self.current < len(self.tokens):
            tok = self.tokens[self.current]
            self.current += 1
            return tok
        return None

    def peek(self) -> Optional[Token]:
        if self.current < len(self.tokens):
            return self.tokens[self.current]
        return None

    def is_next(self, expected_type: TokenType) -> bool:
        assert isinstance(expected_type, TokenType), expected_type
        tok = self.peek()
        assert tok is not None
        return tok.token_type == expected_type

    def expect(self, expected_type: TokenType) -> Token:
        assert isinstance(expected_type, TokenType), expected_type
        tok = self.peek()
        assert tok is not None
        if tok.token_type == expected_type:
            tok = self.advance()
            assert tok is not None
            return tok
        raise SyntaxError(f"Unexpected token {tok}")

    def match(self, token_type: TokenType) -> bool:
        assert isinstance(token_type, TokenType), token_type
        tok = self.peek()
        if not tok or tok.token_type != token_type:
            return False
        self.current += 1
        return True

    def previous_token(self) -> Optional[Token]:
        if self.current - 1 >= 0:
            return self.tokens[self.current - 1]
        return None

    def factor(self) -> Expression:
        return self.unary()

    def exponent(self) -> Expression:
        expr = self.factor()
        while self.match(TokenType.TOK_CARET):
            op = self.previous_token()
            assert op is not None
            right = self.exponent()

            expr = BinOp(op, expr, right)
        return expr

    def modulo(self) -> Expression:
        expr = self.exponent()
        while self.match(TokenType.TOK_MOD):
            op = self.previous_token()
            assert op is not None
            right = self.factor()
            expr = BinOp(op, expr, right)
        return expr

    def term(self) -> Expression:
        expr = self.modulo()
        while self.match(TokenType.TOK_STAR) or self.match(TokenType.TOK_SLASH):
            op = self.previous_token()
            assert op is not None
            right = self.modulo()
            expr = BinOp(op, expr, right)
        return expr

    def unary(self) -> Expression:
        if (
            self.match(TokenType.TOK_NOT)
            or self.match(TokenType.TOK_MINUS)
            or self.match(TokenType.TOK_PLUS)
        ):
            op = self.previous_token()
            assert op is not None
            operand = self.unary()
            return UnaryOp(op, operand)
        return self.primary()

    def primary(self) -> Expression:
        if self.match(TokenType.TOK_INTEGER):
            tok = self.previous_token()
            assert tok is not None
            return Integer(int(tok.lexeme))
        if self.match(TokenType.TOK_FLOAT):
            tok = self.previous_token()
            assert tok is not None
            return Float(float(tok.lexeme))
        if self.match(TokenType.TOK_TRUE):
            tok = self.previous_token()
            assert tok is not None
            return Bool(True)
        if self.match(TokenType.TOK_FALSE):
            tok = self.previous_token()
            assert tok is not None
            return Bool(False)
        if self.match(TokenType.TOK_STRING):
            tok = self.previous_token()
            assert tok is not None
            return String(tok.lexeme[1:-1])
        if self.match(TokenType.TOK_LPAREN):
            expr = self.logical_or()
            if not self.match(TokenType.TOK_RPAREN):
                raise SyntaxError('Error: ")" expected')
            else:
                return Grouping(expr)
        identifier = self.expect(TokenType.TOK_IDENTIFIER)
        return Identifier(identifier.lexeme)

    def expr(self) -> Expression:
        expr = self.term()
        while self.match(TokenType.TOK_PLUS) or self.match(TokenType.TOK_MINUS):
            op = self.previous_token()
            assert op is not None
            right = self.term()
            expr = BinOp(op, expr, right)
        return expr

    def comparison(self) -> Expression:
        expr = self.expr()
        while (
            self.match(TokenType.TOK_GE)
            or self.match(TokenType.TOK_LE)
            or self.match(TokenType.TOK_GT)
            or self.match(TokenType.TOK_LT)
        ):
            op = self.previous_token()
            assert op is not None
            right = self.expr()
            expr = BinOp(op, expr, right)
        return expr

    def equality(self) -> Expression:
        expr = self.comparison()
        while self.match(TokenType.TOK_EQ) or self.match(TokenType.TOK_NE):
            op = self.previous_token()
            assert op is not None
            right = self.comparison()
            expr = BinOp(op, expr, right)
        return expr

    def logical_and(self) -> Expression:
        expr = self.equality()
        while self.match(TokenType.TOK_AND):
            op = self.previous_token()
            assert op is not None
            right = self.equality()
            expr = LogicalOp(op, expr, right)
        return expr

    def logical_or(self) -> Expression:
        expr = self.logical_and()
        while self.match(TokenType.TOK_OR):
            op = self.previous_token()
            assert op is not None
            right = self.logical_and()
            expr = LogicalOp(op, expr, right)
        return expr

    def print_stmt(self) -> PrintStatement:
        if self.match(TokenType.TOK_PRINT):
            return PrintStatement(self.logical_or())
        assert False, "Wrong token type"

    def println_stmt(self) -> PrintlnStatement:
        if self.match(TokenType.TOK_PRINTLN):
            return PrintlnStatement(self.logical_or())
        assert False, "Wrong token type"

    def if_stmt(self) -> IfStatement:
        self.expect(TokenType.TOK_IF)
        test = self.logical_or()
        self.expect(TokenType.TOK_THEN)
        then_stmts = self.stmts()
        else_stmts = None
        if self.is_next(TokenType.TOK_ELSE):
            self.advance()
            else_stmts = self.stmts()
        self.expect(TokenType.TOK_END)
        return IfStatement(test, then_stmts, else_stmts)

    def while_stmt(self) -> WhileStatement:
        self.expect(TokenType.TOK_WHILE)
        test = self.logical_or()
        self.expect(TokenType.TOK_THEN)
        stmts = self.stmts()
        self.expect(TokenType.TOK_END)
        return WhileStatement(test, stmts)

    def for_stmt(self) -> ForStatement:
        self.expect(TokenType.TOK_FOR)
        start = self.stmt()
        assert isinstance(start, Assignment), start
        self.expect(TokenType.TOK_COMMA)
        end = self.logical_or()
        step: Expression = Integer(1)
        if self.match(TokenType.TOK_COMMA):
            step = self.logical_or()
        self.expect(TokenType.TOK_DO)
        stmts = self.stmts()
        self.expect(TokenType.TOK_END)
        return ForStatement(start, end, step, stmts)

    def stmt(self) -> Statement:
        token = self.peek()
        assert token is not None
        match token.token_type:
            case TokenType.TOK_PRINT:
                return self.print_stmt()
            case TokenType.TOK_PRINTLN:
                return self.println_stmt()
            case TokenType.TOK_IF:
                return self.if_stmt()
            case TokenType.TOK_WHILE:
                return self.while_stmt()
            case TokenType.TOK_FOR:
                return self.for_stmt()
            # case TokenType.TOK_FUNC:
            #     return self.func_stmt()
            case _:
                left = self.expr()
                if self.match(TokenType.TOK_ASSIGN):
                    right = self.expr()
                    assert isinstance(left, Identifier), left
                    return Assignment(left, right)
                else:  # TODO: function call
                    pass
        assert False, f"Should not reach here because {token}"

    def stmts(self) -> Statements:
        stmts = []
        while (
            self.current < len(self.tokens)
            and not self.is_next(TokenType.TOK_END)
            and not self.is_next(TokenType.TOK_ELSE)
        ):
            stmts.append(self.stmt())
        return Statements(stmts)

    def parse(self) -> Statements:
        ast = self.stmts()
        return ast

    def __repr__(self) -> str:
        return str(dir(self))

File: zig/src/parser.zig
const std = @import("std");
const tokens = @import("tokens.zig");
const ArrayList = std.ArrayList;
const Expression = @import("model.zig").Expression;
const Node = @import("model.zig").Node;
const Statement = @import("model.zig").Statement;
const Statements = @import("model.zig").Statements;

pub const Parser = struct {
    tokens_list: []tokens.Token,
    current: u32,
    allocator: std.mem.Allocator,

    fn advance(self: *Parser) tokens.Token {
        if (self.current < self.tokens_list.len) {
            self.current += 1;
        }
        return self.tokens_list[self.current - 1];
    }

    fn is_next(self: *Parser, expected_type: tokens.TokenType) bool {
        const token = self.peek();
        if (token == null) return false;
        return token.?.token_type == expected_type;
    }

    fn expect(self: *Parser, expected_type: tokens.TokenType) ?tokens.Token {
        const token = self.peek();
        if (token == null) return null;
        if (token.?.token_type == expected_type) return self.advance();
        unreachable;
    }

    fn peek(self: *Parser) ?tokens.Token {
        if (self.current < self.tokens_list.len) {
            return self.tokens_list[self.current];
        }
        return null;
    }

    fn match_token(self: *Parser, token_type: tokens.TokenType) bool {
        const token = self.peek() orelse return false;
        if (token.token_type == token_type) {
            self.current += 1;
            return true;
        }
        return false;
    }

    fn previous_token(self: *Parser) ?tokens.Token {
        if (self.current >= 1) return self.tokens_list[self.current - 1];
        return null;
    }

    fn exponent(self: *Parser) !*Expression {
        var express = try self.factor();
        while (self.match_token(tokens.TokenType.TokCaret)) {
            const token = self.previous_token().?;
            const fact = try self.exponent();
            const new_express = self.allocator.create(Expression) catch unreachable;
            new_express.* = Expression{ .BinOp = .{ .op = token, .left = express, .right = fact } };
            express = new_express;
        }
        return express;
    }

    fn modulo(self: *Parser) !*Expression {
        var express = try self.exponent();
        while (self.match_token(tokens.TokenType.TokMod)) {
            const token = self.previous_token().?;
            const fact = try self.factor();
            const new_express = self.allocator.create(Expression) catch unreachable;
            new_express.* = Expression{ .BinOp = .{ .op = token, .left = express, .right = fact } };
            express = new_express;
        }
        return express;
    }

    fn term(self: *Parser) !*Expression {
        var express = try self.modulo();
        while (self.match_token(tokens.TokenType.TokStar) or self.match_token(tokens.TokenType.TokSlash)) {
            const token = self.previous_token().?;
            const fact = try self.modulo();
            const new_express = self.allocator.create(Expression) catch unreachable;
            new_express.* = Expression{ .BinOp = .{ .op = token, .left = express, .right = fact } };
            express = new_express;
        }
        return express;
    }

    fn expr(self: *Parser) std.mem.Allocator.Error!*Expression {
        var express = try self.term();
        while (self.match_token(tokens.TokenType.TokPlus) or self.match_token(tokens.TokenType.TokMinus)) {
            const token = self.previous_token().?;
            const term_ = try self.term();
            const new_express = self.allocator.create(Expression) catch unreachable;
            new_express.* = Expression{ .BinOp = .{ .op = token, .left = express, .right = term_ } };
            express = new_express;
        }
        return express;
    }

    fn primary(self: *Parser) !*Expression {
        if (self.match_token(tokens.TokenType.TokInteger)) {
            const token = self.previous_token().?;
            const lexeme_value = std.fmt.parseInt(i32, token.lexeme, 10) catch unreachable;
            const result = self.allocator.create(Expression) catch unreachable;
            result.* = Expression{ .Integer = .{ .value = lexeme_value } };
            return result;
        }
        if (self.match_token(tokens.TokenType.TokFloat)) {
            const token = self.previous_token().?;
            const lexeme_value = std.fmt.parseFloat(f64, token.lexeme) catch unreachable;
            const result = self.allocator.create(Expression) catch unreachable;
            result.* = Expression{ .Float = .{ .value = lexeme_value } };
            return result;
        }
        if (self.match_token(tokens.TokenType.TokTrue)) {
            const result = self.allocator.create(Expression) catch unreachable;
            result.* = Expression{ .Bool = .{ .value = true } };
            return result;
        }
        if (self.match_token(tokens.TokenType.TokFalse)) {
            const result = self.allocator.create(Expression) catch unreachable;
            result.* = Expression{ .Bool = .{ .value = false } };
            return result;
        }
        if (self.match_token(tokens.TokenType.TokString)) {
            const token = self.previous_token().?;
            const len = token.lexeme.len;
            const result = self.allocator.create(Expression) catch unreachable;
            result.* = Expression{ .String = .{ .value = token.lexeme[1 .. len - 1] } };
            return result;
        }
        if (self.match_token(tokens.TokenType.TokLparen)) {
            const express = try self.logical_or();
            if (self.match_token(tokens.TokenType.TokRparen)) {
                const result = self.allocator.create(Expression) catch unreachable;
                result.* = Expression{ .Grouping = .{ .value = express } };
                return result;
            }
        }
        const result = self.allocator.create(Expression) catch unreachable;
        const identifier = self.expect(tokens.TokenType.TokIdentifier).?.lexeme;
        if (self.match_token(tokens.TokenType.TokLparen)) {
            const args = try self.function_params();
            _ = self.expect(tokens.TokenType.TokRparen);
            result.* = Expression{ .FunctionCall = .{ .name = identifier, .args = args } };
        } else {
            result.* = Expression{ .Identifier = .{ .name = identifier } };
        }
        return result;
    }

    fn function_params(self: *Parser) !ArrayList(Expression) {
        var args = ArrayList(Expression).init(self.allocator);
        while (!self.is_next(tokens.TokenType.TokRparen)) {
            try args.append((try self.logical_or()).*);
            if (!self.is_next(tokens.TokenType.TokRparen)) {
                _ = self.expect(tokens.TokenType.TokComma);
            }
        }
        return args;
    }

    fn unary(self: *Parser) !*Expression {
        if (self.match_token(tokens.TokenType.TokNot) or self.match_token(tokens.TokenType.TokMinus) or self.match_token(tokens.TokenType.TokPlus)) {
            const token = self.previous_token().?;
            const un = try self.unary();
            const result = self.allocator.create(Expression) catch unreachable;
            result.* = Expression{ .UnaryOp = .{ .op = token, .exp = un } };
            return result;
        }
        return try self.primary();
    }

    fn factor(self: *Parser) !*Expression {
        return try self.unary();
    }

    fn comparison(self: *Parser) !*Expression {
        var express = try self.expr();
        while (self.match_token(tokens.TokenType.TokGe) or self.match_token(tokens.TokenType.TokLe) or self.match_token(tokens.TokenType.TokGt) or self.match_token(tokens.TokenType.TokLt)) {
            const token = self.previous_token().?;
            const and_ = try self.expr();
            const new_express = self.allocator.create(Expression) catch unreachable;
            new_express.* = Expression{ .BinOp = .{ .op = token, .left = express, .right = and_ } };
            express = new_express;
        }
        return express;
    }

    fn equality(self: *Parser) !*Expression {
        var express = try self.comparison();
        while (self.match_token(tokens.TokenType.TokEq) or self.match_token(tokens.TokenType.TokNe)) {
            const token = self.previous_token().?;
            const and_ = try self.comparison();
            const new_express = self.allocator.create(Expression) catch unreachable;
            new_express.* = Expression{ .BinOp = .{ .op = token, .left = express, .right = and_ } };
            express = new_express;
        }
        return express;
    }

    fn logical_and(self: *Parser) !*Expression {
        var express = try self.equality();
        while (self.match_token(tokens.TokenType.TokAnd)) {
            const token = self.previous_token().?;
            const and_ = try self.logical_and();
            const new_express = self.allocator.create(Expression) catch unreachable;
            new_express.* = Expression{ .LogicalOp = .{ .op = token, .left = express, .right = and_ } };
            express = new_express;
        }
        return express;
    }

    fn logical_or(self: *Parser) !*Expression {
        var express = try self.logical_and();
        while (self.match_token(tokens.TokenType.TokOr)) {
            const token = self.previous_token().?;
            const and_ = try self.logical_and();
            const new_express = self.allocator.create(Expression) catch unreachable;
            new_express.* = Expression{ .LogicalOp = .{ .op = token, .left = express, .right = and_ } };
            express = new_express;
        }
        return express;
    }

    fn if_stmt(self: *Parser) !Statement {
        _ = self.expect(tokens.TokenType.TokIf);
        const test_expr = try self.logical_or();
        _ = self.expect(tokens.TokenType.TokThen);
        const then_stmts = try self.stmts();
        var else_stmts = Statements.init(self.allocator);
        if (self.is_next(tokens.TokenType.TokElse)) {
            _ = self.advance();
            else_stmts = try self.stmts();
        }
        _ = self.expect(tokens.TokenType.TokEnd);
        return Statement{ .IfStatement = .{ .test_expr = test_expr.*, .then_stmts = then_stmts, .else_stmts = else_stmts } };
    }

    fn while_stmt(self: *Parser) !Statement {
        _ = self.expect(tokens.TokenType.TokWhile);
        const test_expr = try self.logical_or();
        _ = self.expect(tokens.TokenType.TokThen);
        const while_stmts = try self.stmts();
        _ = self.expect(tokens.TokenType.TokEnd);
        return Statement{ .While = .{ .test_expr = test_expr.*, .stmts = while_stmts } };
    }

    fn for_stmt(self: *Parser) !Statement {
        _ = self.expect(tokens.TokenType.TokFor);
        const id = try self.primary();
        _ = self.expect(tokens.TokenType.TokAssign);
        const start = try self.logical_or();
        _ = self.expect(tokens.TokenType.TokComma);
        const end = try self.logical_or();
        var step = &Expression{ .Integer = .{ .value = 1 } };
        if (self.match_token(tokens.TokenType.TokComma)) {
            step = try self.logical_or();
        }
        _ = self.expect(tokens.TokenType.TokDo);
        const for_stmts = try self.stmts();
        _ = self.expect(tokens.TokenType.TokEnd);
        return Statement{ .For = .{ .id = id.*, .start = start.*, .end = end.*, .step = step.*, .stmts = for_stmts } };
    }

    fn println_stmt(self: *Parser) !Statement {
        if (self.match_token(tokens.TokenType.TokPrintln)) {
            const express = try self.logical_or();
            return Statement{ .PrintlnStatement = .{ .value = express.* } };
        }
        unreachable;
    }

    fn print_stmt(self: *Parser) !Statement {
        if (self.match_token(tokens.TokenType.TokPrint)) {
            const express = try self.logical_or();
            return Statement{ .PrintStatement = .{ .value = express.* } };
        }
        unreachable;
    }

    fn ret(self: *Parser) !Statement {
        _ = self.expect(tokens.TokenType.TokRet);
        const new_expr = try self.expr();
        return Statement{ .Return = .{ .val = new_expr.* } };
    }

    fn params(self: *Parser) !ArrayList(Statement) {
        var args = ArrayList(Statement).init(self.allocator);
        while (!self.is_next(tokens.TokenType.TokRparen)) {
            const name = self.expect(tokens.TokenType.TokIdentifier).?.lexeme;
            try args.append(Statement{ .Parameter = .{ .name = name } });
            if (!self.is_next(tokens.TokenType.TokRparen)) {
                _ = self.expect(tokens.TokenType.TokComma);
            }
        }
        return args;
    }

    fn function_declaration(self: *Parser) !Statement {
        _ = self.expect(tokens.TokenType.TokFunc);
        const name = self.expect(tokens.TokenType.TokIdentifier).?.lexeme;
        _ = self.expect(tokens.TokenType.TokLparen);
        const args = try self.params();
        _ = self.expect(tokens.TokenType.TokRparen);
        const func_stmts = try self.stmts();
        _ = self.expect(tokens.TokenType.TokEnd);
        return Statement{ .FunctionDeclaration = .{
            .name = name,
            .params = args,
            .stmts = func_stmts,
        } };
    }

    fn local_assignment(self: *Parser) !Statement {
        _ = self.expect(tokens.TokenType.TokLocal);
        const left = try self.expr();
        _ = self.match_token(tokens.TokenType.TokAssign);
        const right = try self.expr();
        return Statement{ .LocalAssignment = .{ .left = left.*, .right = right.* } };
    }

    fn stmt(self: *Parser) !Statement {
        const token = self.peek().?;
        switch (token.token_type) {
            tokens.TokenType.TokPrintln => return self.println_stmt(),
            tokens.TokenType.TokPrint => return self.print_stmt(),
            tokens.TokenType.TokWhile => return self.while_stmt(),
            tokens.TokenType.TokFor => return self.for_stmt(),
            tokens.TokenType.TokIf => return self.if_stmt(),
            tokens.TokenType.TokRet => return self.ret(),
            tokens.TokenType.TokFunc => return self.function_declaration(),
            tokens.TokenType.TokLocal => return self.local_assignment(),
            else => {
                const left = try self.expr();
                if (self.match_token(tokens.TokenType.TokAssign)) {
                    const right = try self.expr();
                    return Statement{ .Assignment = .{ .left = left.*, .right = right.* } };
                }
                return Statement{ .FunctionCall = .{ .expr = left.* } };
            },
        }
    }

    fn stmts(self: *Parser) std.mem.Allocator.Error!Statements {
        var stmts_arr = Statements.init(self.allocator);
        while (self.current < self.tokens_list.len and !self.is_next(tokens.TokenType.TokEnd) and !self.is_next(tokens.TokenType.TokElse)) {
            try stmts_arr.append(try self.stmt());
        }
        return stmts_arr;
    }

    pub fn parse(self: *Parser) !Node {
        return Node{ .Stmts = try self.stmts() };
    }
};
